Assets/Mirror/Runtime/NetworkManager.cs

```diff
@@ -521,122 +521,111 @@ public static void Shutdown()
 
         // ----------------------------- Server Internal Message Handlers  --------------------------------
 
-        internal void OnServerConnectInternal(NetworkMessage netMsg)
+        internal void OnServerConnectInternal(NetworkConnection conn, ConnectMessage connectMsg)
         {
             if (LogFilter.Debug) { Debug.Log("NetworkManager.OnServerConnectInternal"); }
 
             if (networkSceneName != "" && networkSceneName != offlineScene)
             {
-                StringMessage msg = new StringMessage(networkSceneName);
-                netMsg.conn.Send((short)MsgType.Scene, msg);
+                SceneMessage msg = new SceneMessage(networkSceneName);
+                conn.Send(msg);
             }
 
-            OnServerConnect(netMsg.conn);
+            OnServerConnect(conn);
         }
 
-        internal void OnServerDisconnectInternal(NetworkMessage netMsg)
+        internal void OnServerDisconnectInternal(NetworkConnection conn, DisconnectMessage msg)
         {
             if (LogFilter.Debug) { Debug.Log("NetworkManager.OnServerDisconnectInternal"); }
-            OnServerDisconnect(netMsg.conn);
+            OnServerDisconnect(conn);
         }
 
-        internal void OnServerReadyMessageInternal(NetworkMessage netMsg)
+        internal void OnServerReadyMessageInternal(NetworkConnection conn, ReadyMessage msg)
         {
             if (LogFilter.Debug) { Debug.Log("NetworkManager.OnServerReadyMessageInternal"); }
-            OnServerReady(netMsg.conn);
+            OnServerReady(conn);
         }
 
-        internal void OnServerAddPlayerMessageInternal(NetworkMessage netMsg)
+        internal void OnServerAddPlayerMessageInternal(NetworkConnection conn, AddPlayerMessage msg)
         {
             if (LogFilter.Debug) { Debug.Log("NetworkManager.OnServerAddPlayerMessageInternal"); }
 
-            AddPlayerMessage msg = netMsg.ReadMessage<AddPlayerMessage>();
-
             if (msg.value != null && msg.value.Length > 0)
             {
                 // convert payload to extra message and call OnServerAddPlayer
                 // (usually for character selection information)
-                NetworkMessage extraMessage = new NetworkMessage
-                {
-                    reader = new NetworkReader(msg.value),
-                    conn = netMsg.conn
-                };
-                OnServerAddPlayer(netMsg.conn, extraMessage);
+                NetworkReader data = new NetworkReader(msg.value);
+                OnServerAddPlayer(conn, data);
```

Comment:
why not do OnServerAddPlayer(conn, AddPlayerMessage)?
custom add player messages could inherit from AddPlayerMessage and do a simple cast there.

Reply:
I also did not like receiving a reader,  the entire point of this PR is to move away from API like that.

But this one is quite tricky.  You see,  OnServerAddPlayer is not a handler, it is an override.  With handlers I can tell mirror which class I am expecting, which is needed to instantiate it.  But here,  Mirror needs to figure it out. 

I also do not like that there are to OnServerAddPlayer methods depending if there is an extra message or not,  there really needs to be only one.

A possible solution is to do away with the virtual method.   If automatic player add is checked,  we register and send the add player message,  but if it is unchecked,  the user must register a handler and pass me the correct message type.   

I also did not want to do such major surgery in this PR because that is needed to get rid of the NetworkMessage,  but it is not needed for #418 pull request which is what I am trying to get merged right now. This PR is not meant to be merged yet,  it is done because you wanted to see how we could get rid of NetworkMessage. We will need to discuss further about this one method,  it is not that simple.

Reply:
okay so:
AddPlayerMessage already contains a byte[] for the 'extra' part.
so we should just call OnServerAddPlayer(NetworkConnection, AddPlayerMessage) as usual.
those who need extra bytes can do whatever they want with message.value.

for example, in uMMORPG I would just need the character index. so I could convert byte[] to int and call it a day.

inheriting from AddPlayerMessage makes no sense since AddPlayerMessage already already contains a byte[] for extra data.

Reply:
we could definitely do that,  just let the user figure out what to do with the byte[]